### PR TITLE
Enhance messenger doc

### PR DESCRIPTION
**Pull request by ogizanagi**
Backporting some changes made in #10644 + some new changes.

### CODE REVIEW

File: components/messenger.rst [hunk 1]
```diff
@@ -38,11 +38,28 @@ Concepts
...
+**Middleware**:
+   Man-in-the-middle classes, allowing to spy the message and its wrapper (the envelope).
+   Middleware have many usages: log messages, validate messages, start a transaction, ...
+   It can also tweak the envelope, by adding items to it, or even completely replace it
+   or the message inside.
+
+**Envelope**
+   Messenger specific concept, it gives full flexibility inside the message bus,
+   by wrapping the messages into it, allowing to add useful information inside
+   through *envelope items*.
```
Comment (ogizanagi) [inline components/messenger.rst]: Note for 4.2 merge:

```diff
- *envelope items*
+ *envelope stamps*
```

File: components/messenger.rst [hunk 2]
```diff
@@ -38,11 +38,28 @@ Concepts
    something can be a message broker or a third party API for example.
```
Comment (ogizanagi) [inline components/messenger.rst]: Would be great if someone could edit the picture above in order to reflect Envelope usage inside the bus itself IMHO

File: components/messenger.rst [hunk 3]
```diff
@@ -38,11 +38,28 @@ Concepts
...
+   Middleware have many usages: log messages, validate messages, start a transaction, ...
+   It can also tweak the envelope, by adding items to it, or even completely replace it
+   or the message inside.
+
+**Envelope**
+   Messenger specific concept, it gives full flexibility inside the message bus,
+   by wrapping the messages into it, allowing to add useful information inside
+   through *envelope items*.
+
+**Envelope Items**
```
Comment (ogizanagi) [inline components/messenger.rst]: Note for 4.2 merge:

```diff
- **Envelope Items**
+ **Envelope Stamp**
```

File: components/messenger.rst [hunk 4]
```diff
@@ -112,7 +129,7 @@ the ``SerializerConfiguration`` envelope::
         ]))
     );
 
-At the moment, the Symfony Messenger has the following built-in envelopes:
+At the moment, the Symfony Messenger has the following built-in envelope items:
```
Comment (ogizanagi) [inline components/messenger.rst]: Note for 4.2 merge:

```diff
- envelope items
+ envelope stamps
```

File: components/messenger.rst [hunk 5]
```diff
@@ -151,6 +168,12 @@ The above example will forward the message to the next middleware with an additi
 envelope item *if* the message has just been received (i.e. has the `ReceivedMessage` item).
 You can create your own items by implementing :class:`Symfony\\Component\\Messenger\\EnvelopeAwareInterface`.
 
+.. note::
+
+    Any envelope item must be php serializable if going through transport using
```
Comment (ogizanagi) [inline components/messenger.rst]: Note for 4.2 merge:

```diff
- Any envelope item must be php serializable
+ Any envelope item must be serializable using the Symfony Serializer component
```

File: components/messenger.rst [hunk 6]
```diff
@@ -53,9 +70,9 @@ middleware stack. The component comes with a set of middleware that you can use.
 When using the message bus with Symfony's FrameworkBundle, the following middleware
 are configured for you:
 
-#. ``LoggingMiddleware`` (logs the processing of your messages)
-#. ``SendMessageMiddleware`` (enables asynchronous processing)
-#. ``HandleMessageMiddleware`` (calls the registered handle)
+#. :class:`Symfony\\Component\\Messenger\\Middleware\\LoggingMiddleware` (logs the processing of your messages)
```
Comment (ogizanagi) [inline components/messenger.rst]: Note for 4.2 merge: beware of the new FQCNs (already changed)

File: components/messenger.rst [hunk 7]
```diff
@@ -38,11 +38,28 @@ Concepts
...
-   can be a message queue puller or an API endpoint for example.
+   Responsible for retrieving, deserializing and forwarding messages to handler(s).
+   This can be a message queue puller or an API endpoint for example.
 
 **Handler**:
    Responsible for handling messages using the business logic applicable to the messages.
+   Handlers are called by the ``HandleMessageMiddleware`` middleware.
+
+**Middleware**:
+   Man-in-the-middle classes, allowing to spy the message and its wrapper (the envelope).
```
Comment (javiereguiluz) [inline components/messenger.rst]: This phrase sounded a bit weird to me, so I asked on Symfony's Slack and our English native friend Ciaran suggested to reword it as:

```
Intermediary classes, which can inspect the message and its wrapper (the envelope).
```

Comment (ogizanagi) [reply components/messenger.rst]: Thanks! But is `intermediary` really meaningful regarding the concept?

I also found online something we may adapt:

> Middleware (literally the software in the middle) are not core concerns (business logic) of an application. Instead, they are cross cutting concerns applicable throughout the application and affecting the entire message bus.

Comment (ogizanagi) [reply components/messenger.rst]: A bit verbose, but what about:

```rst
   Middleware can access the message and its wrapper (the envelope) while it is
   dispatched through the bus.
   Literally *"the software in the middle"*, those are not about core concerns
   (business logic) of an application. Instead, they are cross cutting concerns
   applicable throughout the application and affecting the entire message bus.
   For instance: logging, validating a message, starting a transaction, ...
   They are also responsible for calling the next middleware in the chain,
   which means they can tweak the envelope, by adding items to it, or even
   replacing it and the message inside, as well as interrupt the middleware chain.
```

?

File: messenger.rst [hunk 8]
```diff
@@ -646,12 +647,18 @@ you can disable them like this:
...
+                        'middleware' => array(
+                            'App\Middleware\MyMiddleware',
+                            'App\Middleware\AnotherMiddleware',
+                        ),
                     ),
                 ),
             ),
         ));
 
+Note that if the service is abstract, a different instance of service will be
```
Comment (xabbuh) [inline messenger.rst]: [...] of the service [...]

File: messenger.rst [hunk 9]
```diff
@@ -729,7 +736,7 @@ factories. Defining such requires a two-step configuration based on Symfony's
 
         // Step 2: an abstract definition that will call the factory with default
         // arguments or the ones provided in the middleware config
-        $container->register('messenger.middleware.doctrine_transaction_middleware', DoctrineTransactionMiddleware::class)
+        $container->register('messenger.middleware.doctrine_transaction', DoctrineTransactionMiddleware::class)
             ->setFactory(array(
                 new Reference('doctrine.orm.messenger.middleware_factory.transaction'),
```
Comment (xabbuh) [inline messenger.rst]: Hasn't the factory been removed?

Comment (ogizanagi) [reply messenger.rst]: Not in 4.1

Comment (xabbuh) [reply messenger.rst]: I thought it was part of DoctrineBundle which doesn't care whether you use the Messenger component in version 4.1 or 4.2. Isn't that the case?

Comment (ogizanagi) [reply messenger.rst]: The factory is in the symfony bridge actually

**Review by xabbuh — APPROVED** [review]

**Comment by javiereguiluz** [comment]
@xabbuh could you please merge this PR? There are a lot of changes that must be done when merging this in `master` and I'm not confident to make all of them right. Thanks!

**Comment by ogizanagi** [comment]
@javiereguiluz: I may provide you a patch to merge in 4.2 tomorrow otherwise.

**Comment by ogizanagi** [comment]
@javiereguiluz : Here are the changes you should expect during merge: https://github.com/ogizanagi/symfony-docs/commit/a6a8617ef625cd29a0036b08d7a1e7717fc37faa (it's probably simpler to c/p the messenger files content to resolve conflicts during merge)

**Comment by xabbuh** [comment]
Thank you @ogizanagi. I merged your PR in the `4.1` branch and made the tweaks you noted in c12928f7db47348274f00dd55edaece20c1a2b3d. I would be happy if you could check that I didn't make any mistakes. :)

**Comment by ogizanagi** [comment]
Awesome! You didn't miss much. I've opened https://github.com/symfony/symfony-docs/pull/10709 for you :)
Thank you.

**Comment by xabbuh** [comment]
@ogizanagi Thank you so much!

**Pull request closed without merging**